src/secp256k1/src/assumptions.h

```diff
@@ -0,0 +1,74 @@
+/**********************************************************************
+ * Copyright (c) 2020 Pieter Wuille                                   *
+ * Distributed under the MIT software license, see the accompanying   *
+ * file COPYING or http://www.opensource.org/licenses/mit-license.php.*
+ **********************************************************************/
+
+#ifndef SECP256K1_ASSUMPTIONS_H
+#define SECP256K1_ASSUMPTIONS_H
+
+#include "util.h"
+
+/* This library, like most software, relies on a number of compiler implementation defined (but not undefined)
+   behaviours. Although the behaviours we require are essentially universal we test them specifically here to
+   reduce the odds of experiencing an unwelcome surprise.
+*/
+
+struct secp256k1_assumption_checker {
```

Comment:
The whole approach here is inherently flawed, because it is based on a false premise: that expressions with undefined behaviour are actually only implementation-defined, and moreover they they are computed in the same way for constant expressions as they are at run-time.

There's no strong reason to believe this should be the case. Obviously it's wrong in theory, but I also mean in practice. Of course constant expressions will be subject to different optimizations than non-constant ones! It's **almost certain** that they will go through different paths in the compiler.

We should remove assumptions about behaviour outside Standard C89, rather than checking them.

Reply:
Implementation-defined behavior and undefined behavior are two distinct concepts in the C language. I believe none of things tested here rely on UB (signed overflow is UB for example, but conversion from unsigned to sign is implementation defined).

Implementation-defined behavior, as far as I know, must mean either raising a signal, or produce a consistent result. In the latter case there is no room for the compiler to evaluate things differently at compile time than at run time.

There is of course no strict guarantee that just because we observe behavior consistent with "conversion from unsigned to signed acts like 2s complement" in these exact inputs that it does for all inputs - but there is only so much that can be checked to warn at compile time. Certain things are still checked at runtime too.

If you believe anything I've said above is wrong, I would appreciate opening an issue so we can fix it.

The rationale for adding these conditions in the first place (see https://github.com/bitcoin-core/secp256k1/pull/798) is because we want to switch to a safegcd-based modular inverse (rather than exponention ladder or libgmp dependency, see https://github.com/bitcoin-core/secp256k1/pull/767) which strongly relies on 2s-complement signed arithmetic. As practically all hardware produced the last decades uses 2s-complement, this is not much of a restriction. This is also reflected in evolution of some standards, e.g. C++20 makes signed shifts well-defined.

In addition, from a process point of view it is very hard to avoid introducing things that inadvertently rely on implementation-defined behavior. ubsan for example (`-fsanitizer=undefined` in GCC and Clang) will warn for UB, and can be configured to warn for a few fully specified but unexpected things like unsigned overflow - but doesn't care about things that are implementation-defined. So even if we wanted to avoid those, it's hard to ensure the existing codebase doesn't already rely on them. We might as well bite the bullet and require essentially universally supported implemented-defined behavior.

Reply:
You're right, on a closer reading, everything here is only implementation-defined. Thankyou for the thorough response!

Part of my confusion was that I had misread line 58.